compiler/packages/babel-plugin-react-compiler/src/Optimization/LowerContextAccess.ts

```diff
@@ -0,0 +1,306 @@
+          contextKeys.has(lvalue.identifier.id)
+        ) {
+          const keys = contextKeys.get(lvalue.identifier.id)!;
+          const selectorFnInstr = emitSelectorFn(fn.env, keys);
+          nextInstructions.push(selectorFnInstr);
+
+          const selectorFn = selectorFnInstr.lvalue;
+          value.args.push(selectorFn);
+        }
+
+        nextInstructions.push(instr);
+      }
+      block.instructions = nextInstructions;
+    }
+    markInstructionIds(fn.body);
+  }
+}
+
+function getContextKeys(value: Destructure): Array<string> | null {
+  const keys = [];
+  const pattern = value.lvalue.pattern;
+
+  switch (pattern.kind) {
+    case 'ArrayPattern': {
+      for (const place of pattern.items) {
+        if (place.kind !== 'Identifier') {
+          return null;
+        }
+
+        if (place.identifier.name === null) {
+          return null;
+        }
+
+        keys.push(place.identifier.name.value);
+      }
+      return keys;
+    }
+
+    case 'ObjectPattern': {
+      for (const place of pattern.properties) {
+        if (
+          place.kind !== 'ObjectProperty' ||
+          place.type !== 'property' ||
+          place.key.kind !== 'identifier'
+        ) {
+          return null;
+        }
+        keys.push(place.key.name);
+      }
+      return keys;
+    }
+  }
+}
+
+function emitPropertyLoad(
+  env: Environment,
+  obj: Place,
+  property: string,
+): {instructions: Array<Instruction>; element: Place} {
+  const loadObj: LoadLocal = {
+    kind: 'LoadLocal',
+    place: obj,
+    loc: GeneratedSource,
+  };
+  const object: Place = {
+    kind: 'Identifier',
+    identifier: makeTemporary(env.nextIdentifierId, GeneratedSource),
+    effect: Effect.Unknown,
+    reactive: false,
+    loc: GeneratedSource,
+  };
+  const loadLocalInstr: Instruction = {
+    lvalue: object,
+    value: loadObj,
+    id: makeInstructionId(0),
+    loc: GeneratedSource,
+  };
+
+  const loadProp: PropertyLoad = {
+    kind: 'PropertyLoad',
+    object,
+    property,
+    loc: GeneratedSource,
+  };
+  const element: Place = {
+    kind: 'Identifier',
+    identifier: makeTemporary(env.nextIdentifierId, GeneratedSource),
+    effect: Effect.Unknown,
+    reactive: false,
+    loc: GeneratedSource,
+  };
+  const loadPropInstr: Instruction = {
+    lvalue: element,
+    value: loadProp,
+    id: makeInstructionId(0),
+    loc: GeneratedSource,
+  };
+  return {
+    instructions: [loadLocalInstr, loadPropInstr],
+    element: element,
+  };
+}
+
+function emitSelectorFn(env: Environment, keys: Array<string>): Instruction {
+  const obj: Place = {
+    kind: 'Identifier',
+    identifier: {
+      id: makeIdentifierId(env.nextIdentifierId),
+      name: makeIdentifierName('c'),
+      mutableRange: {start: makeInstructionId(0), end: makeInstructionId(0)},
+      scope: null,
+      type: makeType(),
+      loc: GeneratedSource,
+    },
+    effect: Effect.Unknown,
+    reactive: false,
+    loc: GeneratedSource,
+  };
```

Comment:
this can be:

const obj = createTemporaryPlace(...);
promoteTemporary(obj);

Reply:
note that it's kind of sketchy to create a new named identifier yourself, it's always best to use promoteTemporary

Reply:
Ah nice! I didn't know about createTemporaryPlace. Fixed